File: agenkit/composition/fallback.py
"""Fallback agent composition pattern."""

from typing import List

from agenkit.interfaces import Agent, Message


class FallbackAgent(Agent):
    """Agent that tries agents in order until one succeeds.

    This implements the Fallback/Retry pattern for reliability.
    """

    def __init__(self, name: str, agents: List[Agent]):
        """Initialize fallback agent.

        Args:
            name: Name of this fallback agent
            agents: List of agents to try in order

        Raises:
            ValueError: If agents list is empty
        """
        if not agents:
            raise ValueError("Fallback agent requires at least one agent")

        self._name = name
        self._agents = agents

    @property
    def name(self) -> str:
        """Return the name of the fallback agent."""
        return self._name

    @property
    def capabilities(self) -> list[str]:
        """Return combined capabilities of all agents."""
        caps_set = set()
        for agent in self._agents:
            caps_set.update(agent.capabilities)
        caps = list(caps_set)
        caps.append("fallback")
        return caps

    async def process(self, message: Message) -> Message:
        """Try each agent in order until one succeeds.

        Args:
            message: Input message

        Returns:
            Response from the first successful agent

        Raises:
            Exception: If all agents fail
        """
        errors = []

        for i, agent in enumerate(self._agents):
            try:
                result = await agent.process(message)

                # Success! Add metadata about which agent was used
                result.metadata["fallback_agent_used"] = agent.name
                result.metadata["fallback_attempt"] = i + 1
                return result

            except Exception as e:
                errors.append(f"agent {i+1} ({agent.name}): {e}")

        # All agents failed
        raise Exception(f"All {len(self._agents)} agents failed: {'; '.join(errors)}")

    def get_agents(self) -> List[Agent]:
        """Return the list of fallback agents."""
        return self._agents

File: agenkit-ts/src/composition/fallback.ts
/**
 * Fallback agent composition pattern.
 *
 * Tries agents in order until one succeeds.
 * This implements the Fallback/Retry pattern for reliability.
 */

import type { Agent, Message } from '../core/interfaces.js';

/**
 * Agent that tries agents in order until one succeeds.
 *
 * This is useful for building fault-tolerant systems where you want
 * to try multiple agents as fallbacks.
 *
 * @example
 * ```typescript
 * const fallback = new FallbackAgent('reliable', [
 *   primaryAgent,
 *   secondaryAgent,
 *   lastResortAgent
 * ]);
 *
 * const result = await fallback.process(message);
 * ```
 */
export class FallbackAgent implements Agent {
  readonly name: string;
  private agents: Agent[];

  /**
   * Create a new fallback agent.
   *
   * @param name - Name of this fallback agent
   * @param agents - List of agents to try in order
   * @throws Error if agents list is empty
   */
  constructor(name: string, agents: Agent[]) {
    if (agents.length === 0) {
      throw new Error('Fallback agent requires at least one agent');
    }

    this.name = name;
    this.agents = agents;
  }

  /**
   * Get combined capabilities of all agents.
   */
  get capabilities(): string[] {
    const capsSet = new Set<string>();

    for (const agent of this.agents) {
      if (agent.capabilities) {
        for (const cap of agent.capabilities) {
          capsSet.add(cap);
        }
      }
    }

    const caps = Array.from(capsSet);
    caps.push('fallback');
    return caps;
  }

  /**
   * Try each agent in order until one succeeds.
   *
   * @param message - Input message
   * @returns Response from the first successful agent
   * @throws Error if all agents fail
   */
  async process(message: Message): Promise<Message> {
    const errors: string[] = [];

    for (let i = 0; i < this.agents.length; i++) {
      const agent = this.agents[i];

      try {
        const result = await agent.process(message);

        // Success! Add metadata about which agent was used
        result.metadata = result.metadata || {};
        result.metadata.fallback_agent_used = agent.name;
        result.metadata.fallback_attempt = i + 1;

        return result;
      } catch (error) {
        const errorMessage =
          error instanceof Error ? error.message : String(error);
        errors.push(`agent ${i + 1} (${agent.name}): ${errorMessage}`);
      }
    }

    // All agents failed
    throw new Error(
      `All ${this.agents.length} agents failed: ${errors.join('; ')}`,
    );
  }

  /**
   * Get the list of fallback agents.
   */
  getAgents(): Agent[] {
    return this.agents;
  }
}
